File: project RAD/distance from origin point/sol.py
import heapq
from typing import List, Tuple

def kClosest(points: List[List[int]], K: int) -> List[List[int]]:
    """
    Finds the K closest points to the origin using a Max-Heap of size K.
    
    Time Complexity: O(N log K)
    Space Complexity: O(K)
    """
    # Max-Heap (simulated using Python's Min-Heap by negating the distance)
    # Heap stores: (-distance_squared, x, y)
    max_heap: List[Tuple[float, int, int]] = []
    
    for x, y in points:
        # Calculate squared distance (no need for sqrt, as D^2 preserves order)
        distance_squared = x**2 + y**2
        
        # We push the negative distance to simulate a Max-Heap (largest distance 
        # becomes the smallest negative number, which Python's Min-Heap handles)
        
        if len(max_heap) < K:
            # 1. Fill the heap until size K
            heapq.heappush(max_heap, (-distance_squared, x, y))
        else:
            # 2. Heap is full, check if the current point is closer than the current farthest (the root)
            # max_heap[0] is the root (the point with the greatest NEGATIVE distance, 
            # meaning the largest POSITIVE distance)
            farthest_distance = -max_heap[0][0] # Convert back to positive for comparison
            
            if distance_squared < farthest_distance:
                # If current point is closer, replace the farthest point
                heapq.heappop(max_heap)
                heapq.heappush(max_heap, (-distance_squared, x, y))
                
    # 3. Extract the final K points from the heap
    result = []
    while max_heap:
        # We only care about the coordinates (x, y)
        neg_dist, x, y = heapq.heappop(max_heap)
        result.append([x, y])
        
    return result

# --- Test Cases ---

if __name__ == "__main__":
    test_cases = [
        ([[1, 3], [-2, 2]], 1, [[-2, 2]]),
        ([[3, 3], [5, -1], [-2, 4]], 2, [[3, 3], [-2, 4]]),
        ([[6, 1], [1, 6], [2, 5]], 2, [[6, 1], [2, 5]]), # Distances: 37, 37, 29. Result should be 29 and one of 37s.
        ([[1, 1], [1, 1]], 1, [[1, 1]]), # Identical points
    ]

    print("--- K Closest Points to Origin Test Results ---")
    
    for points, K, expected_points in test_cases:
        result = kClosest(points, K)
        
        # Sort the results for deterministic comparison (since order doesn't matter)
        sorted_result = sorted(result)
        sorted_expected = sorted(expected_points)
        
        status = "PASSED" if sorted_result == sorted_expected else f"FAILED (Expected: {sorted_expected}, Got: {sorted_result})"
        
        print(f"Points: {points}, K = {K} -> Result: {sorted_result}, Status: {status}")

File: project RAD/distance from origin point/sol.java
import java.util.*;

public class sol {

    public static int[][] kClosest(int[][] points, int K) {
        // Max-Heap (store farthest point on top)
        // Each element = [distanceSquared, x, y]
        PriorityQueue<int[]> maxHeap = new PriorityQueue<>(
                (a, b) -> Integer.compare(b[0], a[0]) // sort in descending order by distance
        );

        for (int[] point : points) {
            int x = point[0];
            int y = point[1];
            int distSq = x * x + y * y;

            // Add to heap
            maxHeap.offer(new int[] { distSq, x, y });

            // If heap size > K, remove the farthest point
            if (maxHeap.size() > K) {
                maxHeap.poll();
            }
        }

        // Prepare result array
        int[][] result = new int[K][2];
        int i = 0;
        while (!maxHeap.isEmpty()) {
            int[] entry = maxHeap.poll();
            result[i][0] = entry[1];
            result[i][1] = entry[2];
            i++;
        }

        return result;
    }

    public static void main(String[] args) {
        // --- Test Cases ---
        int[][][] testPoints = {
                { { 1, 3 }, { -2, 2 } },
                { { 3, 3 }, { 5, -1 }, { -2, 4 } },
                { { 6, 1 }, { 1, 6 }, { 2, 5 } },
                { { 1, 1 }, { 1, 1 } }
        };

        int[] Ks = { 1, 2, 2, 1 };

        System.out.println("--- K Closest Points to Origin Test Results ---");
        for (int t = 0; t < testPoints.length; t++) {
            int[][] result = kClosest(testPoints[t], Ks[t]);

            // Sort for easy comparison
            Arrays.sort(result, (a, b) -> {
                if (a[0] != b[0])
                    return Integer.compare(a[0], b[0]);
                return Integer.compare(a[1], b[1]);
            });

            System.out.print("Input: " + Arrays.deepToString(testPoints[t]));
            System.out.println(", K = " + Ks[t]);
            System.out.println("Result: " + Arrays.deepToString(result));
            System.out.println();
        }
    }
}
